types.go

```diff
@@ -242,20 +242,176 @@ type RaftChanges struct {
 	Leader     raft.ServerID
 }
 
+// VoterEligibility represents whether a node can currently vote,
+// and if it could potentially vote in the future.
+type VoterEligibility struct {
+	currentVoter   bool
+	potentialVoter bool
+}
+
+func (v *VoterEligibility) IsCurrentVoter() bool {
+	return v.currentVoter
+}
+
+func (v *VoterEligibility) IsPotentialVoter() bool {
+	return v.potentialVoter
+}
+
+func (v *VoterEligibility) SetPotentialVoter(isVoter bool) {
+	v.potentialVoter = isVoter
+}
+
+type RaftServerEligibility map[raft.ServerID]*VoterEligibility
+
+// FilterVoters can be used to return all servers that are currently voters,
+// or all servers that are not.
+func (s *RaftServerEligibility) FilterVoters(isCurrentVoter bool) RaftServerEligibility {
+	servers := make(RaftServerEligibility)
+	for id, v := range *s {
+		if v.IsCurrentVoter() == isCurrentVoter {
+			servers[id] = v
+		}
+	}
+	return servers
+}
+
+type CategorizedServers struct {
+	// StaleNonVoters are the IDs of non-voting server nodes in the raft configuration
+	// that are not present in the delegates view of the server nodes should be available
```

Comment:
I'm having trouble parsing this sentence, where does "should be available" fit in?